### PR TITLE
feat: change record key to struct including url and db

### LINKED ISSUE #135: DocumentToSourceRecord: write key as a `Struct` to kafka
## Bug Description

We should write keys as `Structs` with a simple schema consisting of a few fields:

- id (from cloudant)
- url
- db name

This allows users to discriminate events converging on one topic from different cloudant instances. Using a struct guarantees the field order of the serialised key.

Users who don't want or need this can always discard/transform these fields with a simple SMT.

## Environment details

main latest

### PR DESCRIPTION
**Pull request by tomblench**
fixes #135

testing showing new key output:
```sh
./bin/kafka-console-consumer.sh  --topic kafka_test_topic --bootstrap-server localhost:9092  --from-beginning --property print.key=true
...
{"_id":"4826c5d0c7c9eba0babc7a17699405af","cloudant.url":"https://740f610a-33d2-40bd-8f6b-2d4ef06c2140-bluemix.cloudantnosqldb.appdomain.cloud","cloudant.db":"cloudant_source_kafka_test_db"}	{"_rev":"1-967a00dff5e02add41819138abb3284d","_id":"4826c5d0c7c9eba0babc7a17699405af"}
```

### CODE REVIEW

File: src/main/java/com/ibm/cloud/cloudant/kafka/mappers/DocumentToSourceRecord.java [hunk 1]
```diff
@@ -28,8 +31,12 @@
 
 public class DocumentToSourceRecord implements BiFunction<String, ChangesResultItem, SourceRecord> {
 
-    // Record key schema is Map<String, String>
-    private static final Schema RECORD_KEY_SCHEMA = SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.STRING_SCHEMA).build();
+    // Record key schema is {"_id": doc_id, "cloudant.url": url, "cloudant.db": db}
+    private static final Schema RECORD_KEY_SCHEMA = SchemaBuilder.struct()
+            .field(CloudantConst.CLOUDANT_DOC_ID, Schema.STRING_SCHEMA)
+            .field(InterfaceConst.URL, Schema.STRING_SCHEMA)
+            .field(InterfaceConst.DB, Schema.STRING_SCHEMA)
```
Comment (ricellis) [inline src/main/java/com/ibm/cloud/cloudant/kafka/mappers/DocumentToSourceRecord.java:38]: Given that ordering is one of the advantages of the struct I wonder if it makes sense to use either:
* url, db, id
or
* id, db, url

that is to say either "least to most" or "most to least" unique, as it might be more intuitive for human use.

**Review by ricellis — CHANGES_REQUESTED** [review]
I think there also needs to be a documentation update alongside this at https://github.com/IBM/cloudant-kafka-connector/blob/ffb530737e053ca1f360204b28aae3261a0a23a5/README.md?plain=1#L112

I also think we need to update:
https://github.com/IBM/cloudant-kafka-connector/blob/ffb530737e053ca1f360204b28aae3261a0a23a5/src/main/java/com/ibm/cloud/cloudant/kafka/transforms/predicates/IsDesignDocument.java#L51
as it expects a Map not a Struct. The unit tests are very "unit" and make keys of the expected type so they didn't fail with this change.

File: src/main/java/com/ibm/cloud/cloudant/kafka/mappers/DocumentToSourceRecord.java (hunk 1 above)
Comment (tomblench) [reply src/main/java/com/ibm/cloud/cloudant/kafka/mappers/DocumentToSourceRecord.java:38]: Interesting. I first did "url, db, id", then thought about a (premature) optimisation where it's probably best to have the most unique part of the key at the beginning, hence the current order.

On that basis, maybe we should do "id, db, url" which is consistent in being "most to least" and (probably) is a bit more optimal when being processed.

File: README.md [hunk 2]
```diff
@@ -109,9 +109,12 @@ For instance, in the sample configuration files, value schemas are disabled on t
...
-* Keys are produced as `java.util.Map<String, String>` containing an `_id` entry with the original Cloudant document ID.
+* Keys are produced as a `org.apache.kafka.connect.data.Struct` containing:
+  * `_id`: the original Cloudant document ID
+  * `cloudant.db`: the name of the Cloudant database the event originated from
+  * `cloudant.url`: the URL of the Cloudant instance the event originated from.
 * Values are produced as a (schemaless) `java.util.Map<String, Object>`.
-* These types are compatible with the default `org.apache.kafka.connect.json.JsonConverter` and should be compatible with any other converter that can accept a `Map`.
+* These types are compatible with the default `org.apache.kafka.connect.json.JsonConverter` and should be compatible with any other converter that can accept a `Struct` or `Map`.
 * The `schemas.enabled` may be safely used with a `key.converter` if desired.
 * The source connector does not generate schemas for the record values by default. To use `schemas.enable` with the `value.converter` consider using a schema registry or the [`MapToStruct` SMT](docs/smt-reference.md#map-to-struct-conversion).
```
Comment (ricellis) [inline README.md:119]: Scope creep, but can we correct the tense here is the option `schemas.enable` or `schemas.enabled`? one of the lines is wrong 😄

File: src/main/java/com/ibm/cloud/cloudant/kafka/mappers/DocumentToSourceRecord.java [hunk 3]
```diff
@@ -28,8 +31,12 @@
 
 public class DocumentToSourceRecord implements BiFunction<String, ChangesResultItem, SourceRecord> {
 
-    // Record key schema is Map<String, String>
-    private static final Schema RECORD_KEY_SCHEMA = SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.STRING_SCHEMA).build();
+    // Record key schema is {"_id": doc_id, "cloudant.url": url, "cloudant.db": db}
+    public static final Schema RECORD_KEY_SCHEMA = SchemaBuilder.struct()
+            .field(CloudantConst.CLOUDANT_DOC_ID, Schema.STRING_SCHEMA)
+            .field(InterfaceConst.DB, Schema.OPTIONAL_STRING_SCHEMA)
+            .field(InterfaceConst.URL, Schema.OPTIONAL_STRING_SCHEMA)
```
Comment (ricellis) [inline src/main/java/com/ibm/cloud/cloudant/kafka/mappers/DocumentToSourceRecord.java:38]: This is a nice touch, I had been thinking about this also, that making them optional makes it easier to filter them if desired without the need to create a new schema.

File: src/main/java/com/ibm/cloud/cloudant/kafka/mappers/DocumentToSourceRecord.java [hunk 4]
```diff
@@ -79,8 +86,11 @@ public SourceRecord apply(String topic, ChangesResultItem changesResultItem) {
         offsetFunction.apply(changesResultItem.getSeq()),
         topic,
         RECORD_KEY_SCHEMA,
-        Collections.singletonMap(CloudantConst.CLOUDANT_DOC_ID, changesResultItem.getId()),
-        RECORD_VALUE_SCHEMA,
+        new Struct(RECORD_KEY_SCHEMA)
+                .put(CloudantConst.CLOUDANT_DOC_ID, changesResultItem.getId())
+                .put(InterfaceConst.DB, partition.get(InterfaceConst.DB))
+                .put(InterfaceConst.URL, partition.get(InterfaceConst.URL)),
+                RECORD_VALUE_SCHEMA,
```
Comment (ricellis) [inline src/main/java/com/ibm/cloud/cloudant/kafka/mappers/DocumentToSourceRecord.java:93]: This has the wrong indent now
```suggestion
        RECORD_VALUE_SCHEMA,
```

File: src/main/java/com/ibm/cloud/cloudant/kafka/transforms/predicates/IsDesignDocument.java [hunk 5]
```diff
@@ -45,14 +47,13 @@ public ConfigDef config() {
...
-        Map<String, String> key = (Map<String, String>) recordKey;
-        String id = key != null ? key.get(CLOUDANT_DOC_ID) : null;
+        // No need to use message file here, because this purpose gets inserted into the middle of an untranslated Kafka message
+        Requirements.requireSchema(schema, "record key in IsDesignDocument predicate");
+        Requirements.requireStruct(recordKey, "record key in IsDesignDocument predicate");
+        Struct key = (Struct) recordKey;
+        // will throw DataException if not valid
+        key.validate();
+        String id = key.getString(CLOUDANT_DOC_ID);
         if (id == null) {
```
Comment (ricellis) [inline src/main/java/com/ibm/cloud/cloudant/kafka/transforms/predicates/IsDesignDocument.java:57]: I guess this check is no longer necessary as the key schema has a required ID and would throw earlier during `validate` if the id was `null`?

File: src/test/java/com/ibm/cloud/cloudant/kafka/transforms/predicates/IsDesignDocumentTest.java [hunk 6]
```diff
@@ -52,24 +49,24 @@ public void testWrongKeyType() {
 
     @Test(expected = DataException.class)
     public void testKeyNoId() {
-        isDDoc.test(wrapInRecord(RECORD_KEY_SCHEMA, Collections.singletonMap("foo", "bar")));
+        isDDoc.test(wrapInRecord(RECORD_KEY_SCHEMA, new Struct(RECORD_KEY_SCHEMA).put("foo", "bar")));
```
Comment (ricellis) [inline src/test/java/com/ibm/cloud/cloudant/kafka/transforms/predicates/IsDesignDocumentTest.java:52]: I think this test isn't really providing the same "expectation" now. Before it was testing validation in the `isDDoc.test` but my feeling is that now this is throwing a `DataException` simply from the creation of the `Struct` with an invalid field against the new schema.

As such perhaps we should just simplify it to:
```suggestion
        new Struct(RECORD_KEY_SCHEMA).put("foo", "bar");
```
to avoid giving the impression that there is some additional magic happening here. It would mostly be testing the behaviour of the `Struct` class, but does at least validate the key schema a little.

**Review by ricellis — CHANGES_REQUESTED** [review]
nits
